src/binary_space_partition.rs

```diff
@@ -0,0 +1,449 @@
+use std::{fmt, cmp};
+use std::fmt::Write;
+use rand::distributions::WeightedIndex;
+use rand::prelude::*;
+use rand::rngs::StdRng;
+use rand::Rng;
+use serde::Serialize;
+use crate::error::Result;
+
+struct BspLevel {
+    level: Level,
+    map_subsection_min_size: i32,
+    map_subsection_min_room_width: i32,
+    map_subsection_min_room_height: i32,
+}
+
+
+byond_fn!(fn bsp_generate(width, height, hash, map_subsection_min_size, map_subsection_min_room_width, map_subsection_min_room_height) {
+    match bsp_gen(width, height, hash, map_subsection_min_size, map_subsection_min_room_width, map_subsection_min_room_height) {
+        Ok(s) => Some(s),
+        Err(e) => Some(format!("{e}"))
+    }
+});
+fn bsp_gen(width_as_str: &str,
+    height_as_str: &str,
+    hash_as_str: &str,
+    map_subsection_min_size_as_str: &str,
+    map_subsection_min_room_width_as_str: &str,
+    map_subsection_min_room_height_as_str: &str)
+    -> Result<String>{
+    let default_hash: u64 = rand::thread_rng().gen();
+    let width = width_as_str.parse::<i32>()?;
+    let height = height_as_str.parse::<i32>()?;
+    let mut map_subsection_min_size = map_subsection_min_size_as_str.parse::<i32>()?;
+    let map_subsection_min_room_width = map_subsection_min_room_width_as_str.parse::<i32>()?;
+    let map_subsection_min_room_height = map_subsection_min_room_height_as_str.parse::<i32>()?;
+    if map_subsection_min_size < map_subsection_min_room_width || map_subsection_min_size < map_subsection_min_room_height{
+        map_subsection_min_size = cmp::max(map_subsection_min_room_width, map_subsection_min_room_height) + 1
+    }
+    //let seed: &str = Alphanumeric.sample_string(&mut rand::thread_rng(), 32).as_str();
+
+    let mut rng: StdRng = SeedableRng::seed_from_u64(hash_as_str.parse::<usize>()?.try_into().unwrap_or(default_hash));
+
+
+    let level = BspLevel::new(width, height, &mut rng, map_subsection_min_size, map_subsection_min_room_width, map_subsection_min_room_height);
+
+    let mut string = String::new();
+    for room in &level.all_rooms {
+        let serialized = serde_json::to_string(&room).unwrap();
+        let _ = write!(string, "{}", serialized);
+        //println!("serialized = {}", serialized);
+    }
+
+    Ok(format!("{}",serde_json::to_string(&level.all_rooms)?))
```

Comment:
```suggestion
    serde_json::to_string(&level.all_rooms)
```

`serde_json::to_string` already returns a Result<String>, maybe you can just return that?

Reply:
Unfortunately just trying to return the serde_json::to_string gives me
`mismatched types serde_json::Error and error::Error have similar names, but are actually distinct types`
but wrapping it in an Ok()? converts the error into the appropriate error rust-g is expecting